Add listable scope attribute for conversations

PVince81 opened the pull request:
### Related issue

Fixes https://github.com/nextcloud/spreed/issues/1763

### Description

Added ability to set a conversation as listable for regular users and/orguest users from the guest app.
This means those users will be able to search for those rooms and join them even if they haven't joined before.

### Todos
- [x] add new conversation property (mostly copy-pasted from read-only)
   - [x] DB migration for new property
   - [x] backend endpoint to set property
   - [x] settings frontend
   - [x] ~~detect guest app in UI~~ => obsolete ?
- [x] search provider impl
   - [x] impl
- [x] implement search results code based on listable flags
   - [x] show listable only in search results, not regular list
   - [x] filter by scope (regular user vs guest)
   - [x] backend endpoint "/listed-room"
   - [x] frontend
- [x] allow joining listable conversations
- [x] unit tests
   - [x] SystemMessageTest
   - [x] SystemMessage/ListenerTest
   - [x] RoomController, Manager, ParticipantService => cover with integration test ?
   - [x] ~~ListedConversationSearch (if we agree to keep it)~~ removed

### Issues
- [x] BUG: somehow the dialog always shows "Everyone" by default
- [ ] sort out todos in LeftSidebar
   - [x] use ConversationsList ?
   - [ ] sort out "no search results" case, currently nothing appears if no results
   - [x] optional: cancel search when continuing to type? => later in https://github.com/nextcloud/spreed/pull/4742
- [x] implement recognizing guest users (see FIXMEs and TODOs)

### Testing
- [x] verify that cli + defaults there work correctly
- [x] test direct join to private listed room when knowing the URL
- [x] test join via left sidebar
- [ ] test join via global search result
- [x] test with guest users from guest app
- [x] test perms
- [ ] integration tests (see read-only ones)
   - [x] only moderators can change listable property
   - [x] can join room when knowing URL
      - [x] when listed + private (+ system message visible)
      - [x] when listed + public (+ system message visible)
      - [x] when non-listed + public  (+ system message NOT visible)
   - [x] cannot join room when non-listed and private
   - [x] listed rooms of public and private types appear in search results from "/listed-rooms" API, with search term
   - [x] ~~listed rooms of public and private types appear in search results from global search API, with search term~~ => removed

### Open topics

- [x] clarify if flags concept was correct, see https://github.com/nextcloud/spreed/issues/1763#issuecomment-736021636 => adjusted to be a scope, not bitmask any more
- [x] unsure if "all" state is a good idea in regards to extensibility => obsolete
- [x] instead of checkbox, use dropdown for a single-shot choice ? => obsolete
- [x] USER vs USER_SELF_JOINED, see https://github.com/nextcloud/spreed/pull/4706#issuecomment-739883178
- [x] add listable field already in creation dialog ?
- [ ] review discuss the UX concept as the current idea of listable x public/private is already very confusing
- [ ] might need yet another visual indicator in the list to show listed rooms vs non-listed (for moderators to know).
- [x] ~~should we forbid joining listable rooms when calling "/active" endpoint if version used is not v3 ?~~ => no
- [x] password is only asked if not alowed to join a listed room resulting in USER_SELF_JOINED instead of USER, is this ok ?
- [x] should we remove listed conversations from the global search ? => yes, as discussed. they'd cause too much noise there.
- [x] should we force passing in a search term in the "listed-room" API ? or listing all is fine on API level ?

PVince81 commented on appinfo/routes.php:
```diff
@@ -245,6 +245,15 @@
+			'url' => '/api/{apiVersion}/room/{token}/listable',
+			'verb' => 'PUT',
+			'requirements' => [
+				'apiVersion' => 'v(1|2|3)',
```
should this be limited to v3 ? @nickvergessen

nickvergessen replied:
yes

PVince81 commented on lib/Command/Room/Create.php:
```diff
@@ -89,10 +94,21 @@ protected function execute(InputInterface $input, OutputInterface $output): int
+			return 1;
+		}
+
+		if (!in_array($listable, [null, '0', '1', '2', '3'], true)) {
```
seems I messed up the default

nickvergessen replied:
Use the constants (and cast to string?)

PVince81 commented on lib/Migration/Version2100Date20201201102528.php:
```diff
@@ -0,0 +1,60 @@
+use OCP\Migration\SimpleMigrationStep;
+
+/**
+ * Auto-generated migration step: Please modify to your needs!
```
note to self: adjust comment

PVince81 commented on src/services/conversationsService.js:
```diff
@@ -324,6 +324,22 @@ const changeReadOnlyState = async function(token, readOnly) {
+ */
+const changeListable = async function(token, listable) {
+	try {
+		const response = await axios.put(generateOcsUrl('apps/spreed/api/v2', 2) + `room/${token}/listable`, {
```
v3 or whatever we decide for the version

nickvergessen commented on docs/chat.md:
```diff
@@ -147,6 +147,9 @@ Base endpoint is: `/ocs/v2.php/apps/spreed/api/v1`
 * `call_ended` - Call with {user1}, {user2}, {user3}, {user4} and {user5} (Duration 30:23)
 * `read_only_off` - {actor} unlocked the conversation
 * `read_only` - {actor} locked the conversation
+* `listable_participants` - {actor} made the conversation listable only for participants
```
This makes no sense 🙈 i would call it not listable if that is the current state

PVince81 replied:
here with "participant" is implied that the room is only listable by people who already joined them

but I'm fine changing it to "listable_none"

nickvergessen commented on docs/chat.md:
```diff
@@ -147,6 +147,9 @@ Base endpoint is: `/ocs/v2.php/apps/spreed/api/v1`
 * `read_only` - {actor} locked the conversation
+* `listable_participants` - {actor} made the conversation listable only for participants
+* `listable_users` - {actor} made the conversation listable for regular users
+* `listable_all` - {actor} made the conversation listable for everone which includes users and guest users
```
```suggestion
* `listable_all` - {actor} made the conversation listable for everone which includes users and guests
```

nickvergessen commented on docs/constants.md:
```diff
@@ -12,6 +12,11 @@ title: Constants
 
+## Listable scope
+* `0` participants only
+* `1` regular users only, excluding guests
```
Is this guest app users or guests?

PVince81 replied:
as originally clarified in the  ticket https://github.com/nextcloud/spreed/issues/1763#issuecomment-721008640 it's for "guest users from the guest app"

because for public guests (invited by token) we don't have a UI for searching for rooms or switching

nickvergessen commented on lib/Chat/SystemMessage/Listener.php:
```diff
@@ -163,6 +163,24 @@ public static function register(IEventDispatcher $dispatcher): void {
+			$room = $event->getRoom();
+
+			if ($room->getType() === Room::CHANGELOG_CONVERSATION) {
+				return;
```
This case should never be triggered

nickvergessen commented on lib/Command/Room/Update.php:
```diff
@@ -92,6 +98,11 @@ protected function execute(InputInterface $input, OutputInterface $output): int
 			return 1;
 		}
 
+		if (!in_array($listable, [null, '0', '1', '2', '3'], true)) {
```
Further down suggestions only mention 0-2

nickvergessen commented on lib/Migration/Version2100Date20201201102528.php:
```diff
@@ -0,0 +1,60 @@
+			}
+		}
+
+		return $schema;
```
Only return this when the column needs to be added

nickvergessen commented on lib/Room.php:
```diff
@@ -62,6 +62,21 @@ class Room {
+	public const LISTABLE_PARTICIPANTS = 0;
+
+	/**
+	 * Searchable by all regular users and moderators, even when not joined, excluding guest users
```
Guest app users?

nickvergessen commented on lib/Room.php:
```diff
@@ -729,6 +755,42 @@ public function setReadOnly(int $newState): bool {
+			return true;
+		}
+
+		if (!in_array($this->getType(), [self::GROUP_CALL, self::PUBLIC_CALL, self::CHANGELOG_CONVERSATION], true)) {
```
Why changelog? So you can rejoin your own changelog?

nickvergessen commented on lib/Room.php:
```diff
@@ -729,6 +755,42 @@ public function setReadOnly(int $newState): bool {
+			return false;
+		}
+
+		if ($newState < 0 || $newState > 3) {
```
`!in_array` woth the constants? Makes it future saver to extend by searching for the constants

nickvergessen commented on lib/Search/ListedConversationSearch.php:
```diff
@@ -0,0 +1,111 @@
+use OCP\Search\SearchResult;
+use OCP\Search\SearchResultEntry;
+
+class ListedConversationSearch implements IProvider {
```
This should not work via the search from my pov, but the conversation listing/filtering

PVince81 replied:
yes and no

it's still useful to be able to find public conversations with the global search

I have a few new commits that use this provider also for the left side. But that case feels strange.
I'll look into conversation filtering next.

PVince81 replied:
as discussed, we'll remove it

nickvergessen commented on src/components/ConversationSettings/ModerationSettings.vue:
```diff
@@ -21,6 +21,23 @@
+				{{ t('spreed', 'Defines who can find this conversation') }}
+			</div>
+			<div>
+				<label for="moderation_settings_listable_users_conversation_input">{{ t('spreed', 'Listable for') }}</label>
```
Invalid string concatenation?

nickvergessen commented on src/components/ConversationSettings/ModerationSettings.vue:
```diff
@@ -91,25 +108,36 @@
 import SipSettings from './SipSettings'
 
+const listableOptions = [
+	{ value: 2, label: t('spreed', 'Everyone') },
```
src/constants.js 😎

nickvergessen commented on src/components/ConversationSettings/ModerationSettings.vue:
```diff
@@ -91,25 +108,36 @@
 
+const listableOptions = [
+	{ value: 2, label: t('spreed', 'Everyone') },
+	{ value: 1, label: t('spreed', 'Regular users, without guests') },
```
Needs to be clear if its the guest app and to avoid confusion should only say registered/logged in users when guests is not there

PVince81 commented:
I've pushed a few more commits which I had locally which make the search result visible in the left sidebar.

I'm not too happy as it uses the global search provider and it feels wrong.

Tomorrow I'll look into the endpoint that lists joined conversations and see if I can extend it somehow.

nickvergessen commented:
We once also discussed a dedicated view for them? Having it on the normal endpoint would make integration for the mobile clients pretty easy.

PVince81 commented:
- [ ] make sure that other endpoints also allow retrieving listable rooms by token (like getSingleRoom, etc)

PVince81 commented:
> We once also discussed a dedicated view for them? Having it on the normal endpoint would make integration for the mobile clients pretty easy.

yeah, but as a first step we agreed to not handle "guests outside the guest app" for now due to the missing view

PVince81 commented:
Adjusted after review.

Also modified to use a new endpoint "/listed-room" to search and used the `Conversation` component to display results.
Regarding the endpoint, I'm not sure how many fields we want / don't want to be populated, I hoped what I picked is enough for now. A lot of values might be null/zero/default as they are not relevant when the user is not a participant (has not joined the room).

Next up: making it possible to actually join those listable rooms

PVince81 commented:
- [ ] test listable rooms with password...

PVince81 commented:
- [ ] might need yet another visual indicator in the list to show listed rooms vs non-listed.

somehow I dislike the term as it somehow conflicts with "public" vs "non-public". might need to come up with better terms before this is finalized

PVince81 commented:
It is now possible to join listable rooms by clicking on the left sidebar.

Next up: improve the UX after clicking:
- [ ] some JS errors appear
- [ ] it doesn't directly appear in the list and need to wait for the regular refresh => fetch it earlier + scroll to it

PVince81 commented:
after a lot of trial and error I managed to make the transition look clean when clicking on a listable entry: the left sidebar will appear with an entry saying "Joining..." until it gets refreshed.

as an added bonus, the left sidebar now scrolls to the item after joining.

still unresolved issues:
- participant not found in ["updateSessionId" action](https://github.com/nextcloud/spreed/blob/2fc5379f97ec72a7ea5f50519316d33b733b7eed/src/services/participantsService.js#L54)
   - because at the time of joining it's not available yet and might need initialization that only happens further down the road.
   - also somehow it seems the lines are swapped, shouldn't we first set the current participant before calling getCurrentParticipant ?

nickvergessen commented:
> after a lot of trial and error I managed to make the transition look clean when clicking on a listable entry: the left sidebar will appear with an entry saying "Joining..." until it gets refreshed.

The response of the join should contain all the conversation info iirc?

PVince81 commented:
@nickvergessen if the POST to "/active" does, then could try and add it to the store right away.

So far I noticed that the conversation is re-fetched later on which automagically makes it work, somehow.

I still do wonder how non-listable conversations can work correctly in this code path, isn't getParticipantIdentifier() outdated here https://github.com/nextcloud/spreed/blob/2fc5379f97ec72a7ea5f50519316d33b733b7eed/src/services/participantsService.js#L56 ? it might have the value from the previous room, unless I misunderstand something in the "join room dance"

nickvergessen commented on docs/conversation.md:
```diff
@@ -4,6 +4,63 @@
+        ------|------|------------
+        `searchTerm` | string | search term
+
+    - Data:
```
Use this instead of duplicating it:
```suggestion
    - Data: See array definition in `Get user´s conversations`
```

We do already for some other things too and the difference should be 0

nickvergessen commented on lib/Command/Room/Create.php:
```diff
@@ -89,10 +94,26 @@ protected function execute(InputInterface $input, OutputInterface $output): int
 		$owner = $input->getOption('owner');
 		$moderators = $input->getOption('moderator');
 
+		if (!in_array($readOnly, [null, '0', '1'], true)) {
```
constants here too?

nickvergessen commented on lib/Command/Room/Create.php:
```diff
@@ -140,6 +162,14 @@ public function completeOptionValues($optionName, CompletionContext $context) {
 			case 'moderator':
 				return $this->completeParticipantValues($context);
+			case 'readonly':
+				return ['1', '0'];
```
constants

nickvergessen commented on lib/Controller/RoomController.php:
```diff
@@ -210,6 +211,37 @@ public function getRooms(int $noStatusUpdate = 0): DataResponse {
+				// since formatRoom will break early due to having no participant,
+				// we populate the remaining base attributes here
+				$return[] = $this->populateBaseRoomData($roomData, $room, $this->userId);
+				// TODO: should we populate more ?
```
yeah lets stick to the normal output please. Same array as with normal conversations. Otherwise we have follow up regressions with search results not having the same attributes as conversations from the conversation list

PVince81 replied:
we do have the normal output but with zeroed values, as some cannot be populated before we join

I'll check if there's more that we can populate though

PVince81 commented:
participant not found error solved by adding the conversation early, as discussed: https://github.com/nextcloud/spreed/pull/4706/commits/2baed21e41225e41c440ed4821c6111231ab5c6f

nickvergessen commented on lib/Controller/RoomController.php:
```diff
@@ -587,19 +620,12 @@ protected function formatRoomV2andV3(Room $room, ?Participant $currentParticipan
 		$attendee = $currentParticipant->getAttendee();
 		$userId = $attendee->getActorType() === Attendee::ACTOR_USERS ? $attendee->getActorId() : '';
 
+		$roomData = $this->populateBaseRoomData($roomData, $room, $userId);
```
I would just use the `if ($isSIPBridgeRequest) {` block too when:
```php
if ($isSIPBridgeRequest || ($room->getListable() && !$currentParticipant instanceof Participant)) {
```

nickvergessen commented on lib/Controller/RoomController.php:
```diff
@@ -710,6 +736,33 @@ protected function formatRoomV2andV3(Room $room, ?Participant $currentParticipan
+			'readOnly' => $room->getReadOnly(),
+			'listable' => $room->getListable(),
+			'hasCall' => $room->getActiveSince() instanceof \DateTimeInterface,
+			'lobbyState' => $room->getLobbyState(),
```
missing lobbyTimer :boom:

nickvergessen commented on lib/Manager.php:
```diff
@@ -329,6 +335,54 @@ public function getRoomsForUser(string $userId, bool $includeLastMessage = false
+
+		if ($term !== '') {
+			$query->andWhere(
+				$query->expr()->ilike('name', $query->createNamedParameter(
```
```suggestion
				$query->expr()->iLike('name', $query->createNamedParameter(
```

nickvergessen commented on lib/Migration/Version2100Date20201201102528.php:
```diff
@@ -0,0 +1,62 @@
+/**
+ * Add listable column to the rooms table.
+ */
+class Version2100Date20201201102528 extends SimpleMigrationStep {
```
```suggestion
class Version11000Date20201201102528 extends SimpleMigrationStep {
```

nickvergessen commented on lib/Migration/Version2100Date20201201102528.php:
```diff
@@ -0,0 +1,62 @@
+
+			if (!$table->hasColumn('listable')) {
+				$table->addColumn('listable', Type::INTEGER, [
+					'notnull' => true,
```
```suggestion
					'notnull' => false,
```

PVince81 replied:
why have both 0 and null ?
I picked this from read-only which is also not nullable.

nickvergessen replied:
0 and `''` are stored as null by Oracle. Don't ask :speak_no_evil: just allow nullable

nickvergessen commented on lib/Service/ParticipantService.php:
```diff
@@ -161,12 +166,32 @@ public function joinRoom(Room $room, IUser $user, string $password, bool $passed
-			]]);
+			// User joining a group call through listing
+			if (
+				$room->getType() === Room::GROUP_CALL && (
```
or public

PVince81 replied:
public is already handled below, I've left it to "self joined" but still need to test whether it works that way or makes sense

nickvergessen replied:
When you click on a listed conversation you should always be fully added

nickvergessen commented on src/components/LeftSidebar/ConversationsList/Conversation.vue:
```diff
@@ -192,6 +197,11 @@ export default {
 		conversationInformation() {
+			// temporary item while joining
+			if (!this.isSearchResult && !this.item.actorId) {
+				return t('spreed', 'Joining conversation...')
```
```suggestion
				return t('spreed', 'Joining conversation …')
```

nickvergessen commented on src/services/conversationsService.js:
```diff
@@ -110,7 +137,7 @@ const searchPossibleConversations = async function(searchText, token, onlyUsers)
 	}
 
 	try {
-		return await axios.get(generateOcsUrl('core/autocomplete', 2) + `get`, {
+		return axios.get(generateOcsUrl('core/autocomplete', 2) + `get`, {
```
😐 this changes the return from actual data to a promise. doesn't look intended. So better not hide this in a big feature PR?

PVince81 replied:
some leftover from a previous attempt, I'll remove

nickvergessen commented on tests/integration/features/bootstrap/FeatureContext.php:
```diff
@@ -672,6 +676,32 @@ public function userChangesReadOnlyStateOfTheRoom($user, $newState, $identifier,
+	 */
+	public function userChangesListableScopeOfTheRoom($user, $identifier, $newState, $statusCode, $apiVersion = 'v3') {
+		$this->setCurrentUser($user);
+		if ($newState === 'joined') {
```
participants?

PVince81 replied:
=> "none"

nickvergessen commented on lib/Service/ParticipantService.php:
```diff
@@ -161,12 +166,32 @@ public function joinRoom(Room $room, IUser $user, string $password, bool $passed
+				$room->getType() === Room::GROUP_CALL && (
+					// this check should have happened earlier already but let's stay defensive
+					$room->getListable() === Room::LISTABLE_ALL || (
+						$room->getListable() === Room::LISTABLE_USERS && !$this->groupManager->isInGroup($user->getUID(), 'guest_app')
```
use the function from the OCA\Talk\Manager instead, so we don't have multiple places with this magic string?

PVince81 replied:
I couldn't inject Manager without causing an infinite loop of dependencies.

is there an approach for this ?

nickvergessen replied:
Add the function to another service which can be injected in both, but well too much effort. just keep it like this for now.

PVince81 commented:
- [ ] BUG: public link now automatically joins call (not happening on master)

nickvergessen commented on lib/Chat/SystemMessage/Listener.php:
```diff
@@ -200,9 +214,21 @@ public static function register(IEventDispatcher $dispatcher): void {
+				if (
+					// - has joined a file room but not through a public link
+					$userJoinedFileRoom
+					// - has been added by another user
+					|| $userId !== $participant['actorId']
```
So the idea of this was always to not show that you added yourself, when you create a conversation. Maybe you can add this as a comment

PVince81 replied:
I checked and now when creating it also doesn't show up, maybe because when creating you join as "USER_MODERATOR"

nickvergessen commented on lib/Controller/RoomController.php:
```diff
@@ -210,6 +211,33 @@ public function getRooms(int $noStatusUpdate = 0): DataResponse {
+	 */
+	public function getListedRooms(string $searchTerm = ''): DataResponse {
+		$event = new UserEvent($this->userId);
+		$this->dispatcher->dispatch(self::EVENT_BEFORE_LISTED_ROOMS_GET, $event);
```
This should not be needed. Drop it?

We only need it in the other function to check if the changelog needs to be readded and new messages posted

PVince81 replied:
Ok. I just added it for consistency, but if we think no one will use it then I can remove.

nickvergessen commented on src/services/conversationsService.js:
```diff
@@ -88,6 +88,33 @@ const checkTalkVersionHash = function(response) {
+		return response
+	} catch (error) {
+		if (error.response && error.response.status === 503 && !maintenanceWarning) {
+			maintenanceWarning = showError(t('spreed', 'Nextcloud is in maintenance mode, please reload the page'), {
```
This should be handled globally. Feels weird that this search would trigger it but posting a message would not?

nickvergessen replied:
so something for later?

PVince81 replied:
I copy-pasted this from another method. Agreed that this should be global. I'll remove it

PVince81 commented:
- [ ] recheck joining permissions and mode

This is how I think it should be working:
|                            | Regular user | Guest user (guest app) | Public guest (no account) |
|----------------------------|--------------|------------------------|---------------------------|
| Group call, non-listable   | Cannot join  | Cannot join            | Cannot join               |
| Public call, non-listable  | Join USER    | Join USER              | Join USER_SELF_JOINED     |
| Group call, user-listable  | Join USER    | Cannot join            | Cannot join               |
| Public call, user-listable | Join USER    | Join USER              | Join USER_SELF_JOINED     |
| Group call, all-listable   | Join USER    | Join USER              | Cannot join               |
| Public call, all-listable  | Join USER    | Join USER              | Join USER_SELF_JOINED     |

The USER_SELF_JOINED would only ever be for "guests that have no account" (also see https://github.com/nextcloud/spreed/issues/4717 as Talk 10 behaves differently for logged in users)

PVince81 commented:
- [ ] forbid joining any listable rooms if API version is < v3 ?? this way we can say that the new behavior is only in v3

PVince81 commented:
I've pushed some new integration tests.
Some fail because of my misunderstanding of the fact that a USER_SELF_JOINED is not returned as participant of a room. It seems that many existing tests work this around by using the call endpoint instead and counting the peers.

So probably I'll just remove the assertion about "is a participant" and simply rely on the return value of the "/active" call to check if the permission was granted.

nickvergessen commented:
From my POV its:

|                            | Regular user | Guest user (guest app) | Public guest (no account) |
|----------------------------|--------------|------------------------|---------------------------|
| **Group call, non-listable**   | Cannot join  | Cannot join            | Cannot join               |
| **Public call, non-listable**  | USER_SELF_JOINED    | USER_SELF_JOINED               | GUEST     |
| **Group call, user-listable**  | USER    | Cannot join            | Cannot join               |
| **Public call, user-listable** | USER    | USER_SELF_JOINED              | GUEST     |
| **Group call, all-listable**   | USER    | USER              | Cannot join               |
| **Public call, all-listable**  | USER    | USER              | GUEST     |

PVince81 commented:
It took me a while to understand the self joined concept.
Based on my current understanding from https://github.com/nextcloud/spreed/issues/4717 I agree with your proposal as it would keep the behavior consistent until we reach a point where we can adjust it.

- [ ] also make regular user joining "non-listable public" use "USER_SELF_JOINED"

nickvergessen commented:
> It took me a while to understand the self joined concept.

TL;DR: Someone adds you: "USER", you access/join it on your own "USER_SELF_JOINED"

PVince81 commented:
rebased

and I've adjusted the acceptance tests to encode the latter table into test scenarios.

- [ ] troubles with guest user from guest app: https://github.com/nextcloud/guests/issues/505 (the step that adds the user to guest_app doesn't work but [fails silently](https://github.com/nextcloud/guests/issues/504))

PVince81 commented:
I managed to create guest account users (let's call them that for now?) by adding an endpoint to spreedcheats that grabs the GuestManager from the guests app directly.

So now next step is adjusting the actual behavior to match the expected one from the table above.

PVince81 commented:
okay, now the joining-listable-rooms.feature is aligned with the table above and all tests are green

To fix them I also had to set a password for the guest account user and whitelist the spreed app, this is now done in a single test step so it can be done on-demand in scenarios.

nickvergessen commented:
/backport d4ee745 to stable20.1

nickvergessen commented:
/backport d4ee745 to stable20

nickvergessen commented:
Some more tests failing:

> 1) OCA\Talk\Tests\php\Controller\SignalingControllerTest::testLeaveRoomWithOldSession

PVince81 commented:
sharing failures unrelated, also happens on master: https://github.com/nextcloud/spreed/issues/4736

nickvergessen commented:
Not convinced by that. Already see the reports on h1 that the conversation is set to invisible but still shown... But we have more time to find a good word, as long as the api and backend is settled

PVince81 commented:
now also added in the "new conversation dialog":
<img width="222" alt="image" src="https://user-images.githubusercontent.com/277525/101791386-c4586480-3b03-11eb-9371-abf0419361d8.png">

I've reused the same component there but the saving happens later at creation, it needed an extra request.
Not sure if we should bundle "listable" in the initial creation request instead.

PVince81 commented:
While developing this I was slightly bothered that search requests did not get cancelled while typing.
Since this is slightly off topic for this PR I pushed it to another one https://github.com/nextcloud/spreed/pull/4742 which depends on this one here due to the additional search query for listable conversations.

PVince81 commented:
I've squashed some commits together that belonged together and resolved the DCO issue.

nickvergessen commented on lib/Migration/Version11000Date20201201102528.php:
```diff
@@ -0,0 +1,62 @@
+namespace OCA\Talk\Migration;
+
+use Closure;
+use Doctrine\DBAL\Types\Type;
```
```suggestion
use Doctrine\DBAL\Types\Types;
```

PVince81 commented:
- [ ] delete global search code as discussed with @nickvergessen

nickvergessen commented on lib/Migration/Version11000Date20201201102528.php:
```diff
@@ -0,0 +1,62 @@
+			$table = $schema->getTable('talk_rooms');
+
+			if (!$table->hasColumn('listable')) {
+				$table->addColumn('listable', Type::INTEGER, [
+					'notnull' => false,
+					'length' => 6,
+					'default' => 0,
```
```suggestion
				$table->addColumn('listable', Types::SMALLINT, [
					'notnull' => false,
					'default' => 0,
					'unsigned' => true,
```

nickvergessen commented on lib/Migration/Version11000Date20201201102528.php:
```diff
@@ -0,0 +1,62 @@
+					'notnull' => false,
+					'length' => 6,
+					'default' => 0,
+				]);
```
We need an index on the column, otherwise the search will die at some point.

nickvergessen commented on lib/Service/ParticipantService.php:
```diff
@@ -166,12 +171,34 @@ public function joinRoom(Room $room, IUser $user, string $password, bool $passed
+					$room->getListable() === Room::LISTABLE_ALL || (
+						$room->getListable() === Room::LISTABLE_USERS &&
+						// queried here to avoid loop deps
+						!\OC::$server->query(\OCA\Talk\Manager::class)->isGuestUser($user->getUID())
```
```suggestion
						!\OC::$server->get(\OCA\Talk\Manager::class)->isGuestUser($user->getUID())
```

nickvergessen commented on src/components/ConversationSettings/ListableSettings.vue:
```diff
@@ -0,0 +1,131 @@
+		</div>
+		<div>
+			<label for="moderation_settings_listable_conversation_input">{{ t('spreed', 'Visible for') }}</label>
+			<Multiselect id="moderation_settings_listable_conversation_input"
+				v-model="listable"
+				:options="listableOptions"
+				:placeholder="t('spreed', 'Visible for')"
+				label="label"
+				track-by="value"
+				:disabled="isListableLoading"
+				aria-describedby="moderation_settings_listable_conversation_hint"
+				@input="saveListable" />
```
Needs to be a bit wider
![Bildschirmfoto von 2020-12-11 09-41-07](https://user-images.githubusercontent.com/213943/101881693-109fb500-3b95-11eb-8c98-fc7d101530c9.png)

nickvergessen commented on src/components/ConversationSettings/ModerationSettings.vue:
```diff
@@ -21,6 +21,7 @@
 
 <template>
 	<div>
+		<ListableSettings :token="token" />
```
Currently the settings block is captioned with `Meeting settings` which is weird.
Only the lobby refers to meetings. "listable" and "locking" should be `Conversation settings` instead of `Meeting settings`

nickvergessen commented on src/services/conversationsService.js:
```diff
@@ -324,9 +343,26 @@ const changeReadOnlyState = async function(token, readOnly) {
+		})
+		return response
+	} catch (error) {
+		console.debug('Error while updating listable scope: ', error)
```
Since you catch here, `await this.$store.dispatch('setListable', {` doesn't catch anything anymore and does not `showError`, but always shows the success message.
can be tested by putting
```php
return new DataResponse([], Http::STATUS_BAD_REQUEST);
```
In the first line of the function in the RoomController.

From my pov the services should never catch.

nickvergessen commented on tests/integration/features/bootstrap/FeatureContext.php:
```diff
@@ -121,6 +133,81 @@ public function tearDown() {
+	}
+
+	/**
+	 * @Then /^user "([^"]*)" cannot find any listed rooms with (\d+)(?: \((v(1|2|3))\))?$/
+	 *
+	 * @param string $user
+	 * @param int $statusCode
+	 * @param string $apiVersion
+	 */
+	public function userCannotFindAnyListedRoomsWithStatus($user, $statusCode, $apiVersion = 'v1') {
```
```suggestion
	 * @Then /^user "([^"]*)" cannot find any listed rooms with (\d+)(?: \((v(3))\))?$/
	 *
	 * @param string $user
	 * @param int $statusCode
	 * @param string $apiVersion
	 */
	public function userCannotFindAnyListedRoomsWithStatus($user, $statusCode, $apiVersion = 'v3') {
```

nickvergessen commented on tests/integration/features/bootstrap/FeatureContext.php:
```diff
@@ -121,6 +133,81 @@ public function tearDown() {
+	}
+
+	/**
+	 * @Then /^user "([^"]*)" cannot find any listed rooms with term "([^"]*)"(?: \((v(1|2|3))\))?$/
+	 *
+	 * @param string $user
+	 * @param string $term
+	 * @param string $apiVersion
+	 */
+	public function userCannotFindAnyListedRoomsWithTerm($user, $term, $apiVersion = 'v1') {
```
```suggestion
	 * @Then /^user "([^"]*)" cannot find any listed rooms with term "([^"]*)"(?: \((v(3))\))?$/
	 *
	 * @param string $user
	 * @param string $term
	 * @param string $apiVersion
	 */
	public function userCannotFindAnyListedRoomsWithTerm($user, $term, $apiVersion = 'v3') {
```

nickvergessen replied:
and so on

nickvergessen commented on tests/integration/features/bootstrap/FeatureContext.php:
```diff
@@ -793,6 +903,36 @@ public function userChangesReadOnlyStateOfTheRoom($user, $newState, $identifier,
 	}
 
+	/**
+	 * @Then /^user "([^"]*)" allows listing room "([^"]*)" for "(none|users|all|\d+)" with (\d+)(?: \((v(1|2|3))\))?$/
```
```suggestion
	 * @Then /^user "([^"]*)" allows listing room "([^"]*)" for "(none|users|all|\d+)" with (\d+)(?: \((v(3))\))?$/
```

PVince81 commented:
- [ ] add missing check for guests in resolveRoomDisplayName (see FIXME in lib/Manager.php)

PVince81 commented:
DCO again ? :-(

nickvergessen commented on lib/Manager.php:
```diff
@@ -447,8 +511,20 @@ public function getRoomForUserByToken(string $token, ?string $userId, bool $incl
+
+			// never joined before but found in listing
+			$listable = null;
+			if (isset($row['listable'])) {
```
Why isset? The query will always select it?

nickvergessen commented:
> DCO again ? :-(

The revert is the problem.

PVince81 commented:
removed redundant isset, fixed DCO issue, rebased

expecting green now :-)

nickvergessen commented:
Version bump got lost on rebase because master already is `11.0.0-dev.9`

nickvergessen commented on docs/conversation.md:
```diff
@@ -4,6 +4,63 @@
+    - Data: See array definition in `Get user´s conversations`
+        Array of conversations, each conversation has at least:
+
+        field | type | API | Description
+        ------|------|-----|------------
+        `token` | string | * | Token identifier of the conversation which is used for further interaction
+        `type` | int | * | See list of conversation types in the [constants list](constants.md#Conversation-types)
+        `name` | string | * | Name of the conversation (can also be empty)
+        `displayName` | string | * | `name` if non empty, otherwise it falls back to a list of participants
+        `participantType` | int | * | Permissions level of the current user
+        `attendeeId` | int | v3 | Unique attendee id
+        `attendeePin` | string | v3 | Unique dial-in authentication code for this user, when the conversation has SIP enabled (see `sipEnabled` attribute)
+        `actorType` | string | v3 | Currently known `users|guests|emails|groups`
+        `actorId` | string | v3 | The unique identifier for the given actor type
+        `participantInCall` | bool | 🏴 v1 | Flag if the current user is in the call (deprecated, use `participantFlags` instead)
+        `participantFlags` | int | * | Flags of the current user (only available with `in-call-flags` capability)
+        `readOnly` | int | * | Read-only state for the current user (only available with `read-only-rooms` capability)
+        `listable` | int | * | Listable scope for the room (only available with `listable-rooms` capability)
+        `count` | int | 🏴 v1 | **Deprecated:** ~~Number of active users~~ - always returns `0`
+        `numGuests` | int | 🏴 v1 | Number of active guests
+        `lastPing` | int | * | Timestamp of the last ping of the current user (should be used for sorting)
+        `sessionId` | string | * | `'0'` if not connected, otherwise a 512 character long string
+        `hasPassword` | bool | * | Flag if the conversation has a password
+        `hasCall` | bool | * | Flag if the conversation has an active call
+        `canStartCall` | bool | * | Flag if the user can start a new call in this conversation (joining is always possible) (only available with `start-call-flag` capability)
+        `canDeleteConversation` | bool | 🆕 v2 | Flag if the user can delete the conversation for everyone (not possible without moderator permissions or in one-to-one conversations)
+        `canLeaveConversation` | bool | 🆕 v2 | Flag if the user can leave the conversation (not possible for the last user with moderator permissions)
+        `lastActivity` | int | * | Timestamp of the last activity in the conversation, in seconds and UTC time zone
+        `isFavorite` | bool | * | Flag if the conversation is favorited by the user
+        `notificationLevel` | int | * | The notification level for the user (one of `Participant::NOTIFY_*` (1-3))
+        `lobbyState` | int | * | Webinary lobby restriction (0-1), if the participant is a moderator they can always join the conversation (only available with `webinary-lobby` capability)
+        `lobbyTimer` | int | * | Timestamp when the lobby will be automatically disabled (only available with `webinary-lobby` capability)
+        `sipEnabled` | int | v3 | SIP enable status (0-1)
+        `canEnableSIP` | int | v3 | Whether the given user can enable SIP for this conversation. Note that when the token is not-numeric only, SIP can not be enabled even if the user is permitted and a moderator of the conversation
+        `unreadMessages` | int | * | Number of unread chat messages in the conversation (only available with `chat-v2` capability)
+        `unreadMention` | bool | * | Flag if the user was mentioned since their last visit
+        `lastReadMessage` | int | * | ID of the last read message in a room (only available with `chat-read-marker` capability)
+        `lastMessage` | message | * | Last message in a conversation if available, otherwise empty
+        `objectType` | string | * | The type of object that the conversation is associated with; "share:password" if the conversation is used to request a password for a share, otherwise empty
+        `objectId` | string | * | Share token if "objectType" is "share:password", otherwise empty
```
Drop please

PVince81 commented:
as discussed, I've adjusted the listable dropdown width and tested with the dyslexia font:
<img width="305" alt="image" src="https://user-images.githubusercontent.com/277525/102072719-8f4d5a00-3e02-11eb-9358-9651604e0a3d.png">

PVince81 commented:
we didn't backport the conversation settings: https://github.com/nextcloud/spreed/pull/4576

unless critical, I'd rather not backport these huge changes

nickvergessen commented:
we shouldn't backport this at all...

PVince81 commented:
ah, seems you asked the bot to backport a single commit about the ctrl+c... not sure why there would be conflict, maybe the bot tried to backport all

I can do this manually

PVince81 commented:
here we go: https://github.com/nextcloud/spreed/pull/4759

nickvergessen commented:
Ah, no the problem is you rebased and the sha of the commit didn't exist anymore.

PVince81 commented:
this PR will remove the guest manager cheat: https://github.com/nextcloud/spreed/pull/4764
